### ISSUE TITLE
MacOS can't specify accellerator as hvf with qemuargs

### ISSUE THREAD

**Issue by paul-nelson-baker** [vagrant-libvirt/vagrant-libvirt#1337]
### Steps to reproduce
1. I followed the relevant steps from [Issue 497](https://github.com/vagrant-libvirt/vagrant-libvirt/issues/497#issuecomment-331226071) to compile and install the gem into vagrant's embedded ruby.
2. I set my environment variable for default provider to libvirt
    ```bash
    export VAGRANT_DEFAULT_PROVIDER=libvirt
    ```
3. Execute `vagrant up`

### Expected behaviour
Should start up an archlinux VM via libvirt/QEMU because it is attempting to use kvm instead of hvf even with explicit qemu args.

### Actual behaviour
Vagrant fails:
```
Bringing machine 'default' up with 'libvirt' provider...
==> default: Checking if box 'archlinux/archlinux' version '20210815.31636' is up to date...
==> default: Creating image (snapshot of base box volume).
==> default: Creating domain with the following settings...
==> default:  -- Name:              its-still-docker_default
==> default:  -- Description:       Source: /Users/pnc006/amfam-code/its-still-docker/Vagrantfile
==> default:  -- Domain type:       kvm
==> default:  -- Cpus:              1
==> default:  -- Feature:           acpi
==> default:  -- Feature:           apic
==> default:  -- Feature:           pae
==> default:  -- Clock offset:      utc
==> default:  -- Memory:            512M
==> default:  -- Management MAC:
==> default:  -- Loader:
==> default:  -- Nvram:
==> default:  -- Base box:          archlinux/archlinux
==> default:  -- Storage pool:      default
==> default:  -- Image():     /Users/pnc006/.local/share/libvirt/images/its-still-docker_default.img, 20G
==> default:  -- Disk driver opts:  cache='default'
==> default:  -- Kernel:
==> default:  -- Initrd:
==> default:  -- Graphics Type:     vnc
==> default:  -- Graphics Port:     -1
==> default:  -- Graphics IP:       127.0.0.1
==> default:  -- Graphics Password: Not defined
==> default:  -- Video Type:        cirrus
==> default:  -- Video VRAM:        9216
==> default:  -- Sound Type:
==> default:  -- Keymap:            en-us
==> default:  -- TPM Backend:       passthrough
==> default:  -- TPM Path:
==> default:  -- INPUT:             type=mouse, bus=ps2
==> default:  -- Command line args:
==> default:     -> value=-machine,
==> default:     -> value=accel=hvf,
Error while creating domain: Error saving the server: Call to virDomainDefineXML failed: invalid argument: could not get preferred machine for /usr/local/bin/qemu-system-x86_64 type=kvm
```

The relevant portion of the error:
```
failed: invalid argument: could not get preferred machine for /usr/local/bin/qemu-system-x86_64 type=kvm
```

### System configuration
**OS/Distro version:**:
- MacOS Catalina 10.15.7

**Libvirt version:**
- libvirt: stable 7.7.0 (bottled), HEAD
- QEMU emulator version 6.1.0

**Output of `vagrant version; vagrant plugin list`:**
```
Installed Version: 2.2.18
Latest Version: 2.2.18
 
You're running an up-to-date version of Vagrant!
vagrant-libvirt (0.5.3, global)
  - Version Constraint: > 0
```

**Attach Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`**
[vagrant-up.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/7107768/vagrant-up.log)

**A Vagrantfile to reproduce the issue:**
```ruby
Vagrant.configure("2") do |config|
  config.vm.synced_folder ".", "/vagrant", disabled: true
  config.vm.box = "archlinux/archlinux"
  config.vm.provider :libvirt do |libvirt|
    libvirt.uri = "qemu:///session"
    libvirt.socket = "#{Dir.home}/.cache/libvirt/libvirt-sock"
    libvirt.qemuargs :value => '-machine'
    libvirt.qemuargs :value => 'accel=hvf'
  end
end
```

**Are you using upstream vagrant package or your distros package?**
Not this time, just the currently installed version available to homebrew

**Comment by abbbi** [comment]
Unfortunately the debug output doesnt include the generated virtual machine XML, could check if VM is created and post "virsh dumpxml <machine>" output?

I could imagine the "machine_type" option could help here (see README).

**Comment by paul-nelson-baker** [comment]
Sorry for the late reply, my previous sprint got overburdened and then I was moving in my personal life.

I'm not sure exactly what the order of operations is for libvirt under the hood, but I tried setting it a number of ways. None of these seemed to actually get picked up because it still showed KVM as the type:
* `libvirt.machine_type = "hvf"`
* `libvirt.machine_type = "accel=hvf"`
* `libvirt.machine_type = "pc,accel=hvf"`

I tried to get the dumpxml output, but I think something is missing because this didn't work either:
```
$ virsh --connect "qemu+unix:///session?socket=${HOME}/.cache/libvirt/libvirt-sock" dumpxml
error: command 'dumpxml' requires <domain> option

$ virsh --connect "qemu+unix:///session?socket=${HOME}/.cache/libvirt/libvirt-sock" list
 Id   Name   State
--------------------
```

So no domain is created by vagrant/libvirt, but I expect this is supposed to do this automatically since no other instructions dictate this must be done before hand. Where should I go to find the generated XML?

**Comment by electrofelix** [comment]
It looks like the error occurs during the domain creation. Since that failed there won't be any domain to dump the XML from.

It looks like what you really want to set for the qemuargs is `-accel=hvf` rather than trying to compete with the `-machine` arg and the order of options passed by libvirt:
```
    libvirt.qemuargs :value => '-accel'
    libvirt.qemuargs :value => 'hvf'
```
It may not work, so it might be worth looking at the output of `qemu-system-x86_64 --help` for `-machine` and also look at `qemu-system-x86_64 -machine help` because I think when you set `-machine` you need to specify one of the supported machines and then add `accel` as a property. I don't think you can just ask it to set the accel property by itself via that arg.

To get better debug on the XML being sent to libvirt in case that doesn't work, are you able to build and install the latest plugin code from source using ruby? It contains some additional debug that will display the XML that is being sent to libvirt so we can see what it is passing to libvirt.
Basic steps would be:
```bash
git clone https://github.com/vagrant-libvirt/vagrant-libvirt
cd vagrant-libvirt
bundle config set --local path `pwd`/vendor
bundle install
bundle exec rake build
```

You'll still need to use the additional steps you followed to install the resulting plugin package.

**Comment by electrofelix** [comment]
@paul-nelson-baker were you able to adjust the commands as I suggested or install any of the release since 0.6.0 which will capture the XML sent as part of the debug log?

**Comment by atreides322** [comment]
I was running into this same issue. I was unable to install vagrant-libvirt from source, but I have the newest released version (0.7.0) installed and am able to see the domain XML being generated. It looks like the issue stems from:
```xml
<domain type='kvm' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>
```
I came across #1017, where I go the idea to set `libvirt.driver = 'qemu'`. The Vagrantfile now looks like:
```ruby
Vagrant.configure('2') do |config|
  config.vm.synced_folder '.', '/vagrant', disabled: true
  config.vm.box = 'generic/ubuntu2004'
  config.vm.provider :libvirt do |libvirt|
    libvirt.driver = 'qemu'
    libvirt.uri = 'qemu:///session'
    libvirt.socket = '#{Dir.home}/.cache/libvirt/libvirt-sock'
    libvirt.machine_type = 'pc'
    libvirt.qemuargs :value => '-accel'
    libvirt.qemuargs :value => 'hvf'
  end
end
```
The domain now gets the correct type:
```xml
<domain type='qemu' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>
```
But now I am getting the same `Call to virConnectNumOfNetworks failed` error as #1017.
```
/Users/atreides/.vagrant.d/gems/2.7.4/gems/vagrant-libvirt-0.7.0/lib/vagrant-libvirt/util/network_util.rb:144:in `list_networks': Call to virConnectNumOfNetworks failed: this function is not supported by the connection driver: virConnectNumOfNetworks (Libvirt::RetrieveError)
```

**Comment by electrofelix** [comment]
@atreides322 could you share the full qemu command line that is executed here in this issue?

from the settings:
```
    libvirt.driver = 'qemu'
    libvirt.machine_type = 'pc'
```
and needing to provide:
```
    libvirt.qemuargs :value => '-accel'
    libvirt.qemuargs :value => 'hvf'
```

I think we need to add an entry explicitly for a `hvf` driver that then sets the correct combination of settings for hvf.

**Comment by atreides322** [comment]
@electrofelix The networking issue, which caused me to move the discussion to the other ticket, was actually masking another issue. If I use:
```ruby
    libvirt.driver = 'qemu'
    libvirt.machine_type = 'pc'
```
and 
```ruby
    libvirt.qemuargs :value => '-accel'
    libvirt.qemuargs :value => 'hvf'
```
and try to manually start the domain, I get the following error:
```
error: Failed to start domain 'libvirt-hvf_default'
error: internal error: process exited while connecting to monitor: 2021-11-21T14:53:38.061645Z qemu-system-x86_64: The -accel and "-machine accel=" options are incompatible
```
The correct arguments appear to be:
```ruby
    libvirt.qemuargs :value => '-machine'
    libvirt.qemuargs :value => 'accel=hvf'
```
Also, it looks as though specifying the machine type is unnecessary because the version of libvirt installed by Homebrew uses `pc-i440fx-6.1` by default, which `pc` is an alias for.

I am not sure which qemu command you are looking for. My final, "working" Vagrantfile looks like:
```ruby
Vagrant.configure('2') do |config|
  config.vm.synced_folder '.', '/vagrant', disabled: true
  config.vm.box = 'generic/ubuntu2004'
  config.vm.provider :libvirt do |libvirt|
    libvirt.driver = 'qemu'
    libvirt.uri = 'qemu:///session'
    libvirt.socket = '#{Dir.home}/.cache/libvirt/libvirt-sock'
    libvirt.qemuargs :value => '-machine'
    libvirt.qemuargs :value => 'accel=hvf'
    libvirt.qemuargs :value => '-netdev'
    libvirt.qemuargs :value => 'user,id=n0,hostfwd=tcp::2222-:22'
    libvirt.qemuargs :value => '-device'
    libvirt.qemuargs :value => 'virtio-net-pci,netdev=n0'
  end
end
```
And if it helps, I've attached of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`. Maybe the contains the information you're requesting.
[vagrant.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/7576499/vagrant.log)

**Comment by atreides322** [comment]
Due to a recent change in libvirt 7.10.0, providing `-machine accel=hvf` no longer works:

https://gitlab.com/libvirt/libvirt/-/issues/147#note_755379369

**Comment by andreabolognani** [comment]
Native support for Hypervisor.framework acceleration has now been merged into libvirt, and will be part of the upcoming 8.1.0 release (scheduled for early March).

This should make it possible to finally implement proper support for it in vagrant-libvirt, e.g. `libvirt.driver = 'hvf'`. The `libvirt.qemuargs` method of enabling hardware acceleration doesn't work as of 7.10.0 and will not work going forward.

**Comment by electrofelix** [comment]
If anyone is who has got vagrant+vagrant-libvirt working on MacOS would like to test out the gem attached to #1742 it might be possible to resolve this issue.
